File: server/app/schemas/grouping.py
"""DRG 入组相关 Pydantic Schema。参照 plans/03_api_interfaces.md §4。"""

from __future__ import annotations

from typing import Any

from app.schemas.common import CamelModel


class GroupingExecuteRequest(CamelModel):
    case_id: str
    rule_version_id: str | None = None


class BatchGroupingRequest(CamelModel):
    case_ids: list[str]
    rule_version_id: str | None = None


class EvidenceItem(CamelModel):
    step: int
    type: str
    description: str
    matched_code: Any | None = None
    matched_rule: str | None = None
    cc_level: str | None = None
    excluded_by: list[str] | None = None
    excluded: bool | None = None


class CandidateRule(CamelModel):
    adrg: str | None = None
    drg: str | None = None
    name: str | None = None
    reason: str | None = None
    hit: bool | None = None


class GroupingTaskResponse(CamelModel):
    task_id: str
    status: str
    started_at: Any | None = None


class GroupingResultDetail(CamelModel):
    mdc: dict | None = None
    adrg: dict | None = None
    drg: dict | None = None
    complication: str | None = None
    evidence: list[dict] = []
    explanation: str | None = None
    candidate_rules: list[dict] = []
    warnings: list[str] = []


class GroupingResultResponse(CamelModel):
    task_id: str
    status: str
    case_id: str | None = None
    rule_version: str | None = None
    started_at: Any | None = None
    finished_at: Any | None = None
    duration_ms: int | None = None
    result: GroupingResultDetail | None = None
    error: dict | None = None
    input_snapshot: dict | None = None

File: web/src/types/grouping.ts
import type { TaskStatus } from './api';

export interface GroupingExecuteRequest {
  caseId: string;
  ruleVersionId: string;
}

export interface GroupingExecuteResponse {
  taskId: string;
  status: TaskStatus;
  startedAt: string;
}

export interface GroupingCodeName {
  code: string;
  name: string;
}

export interface EvidenceItem {
  step: number;
  type: string;
  description: string;
  matchedCode?: string;
  matchedRule?: string;
  ccLevel?: 'MCC' | 'CC' | 'NONE';
  excludedBy?: string[];
  excluded?: boolean;
}

export interface CandidateRule {
  adrg: string;
  drg: string;
  name: string;
  reason: string;
}

export interface GroupingResult {
  mdc: GroupingCodeName;
  adrg: GroupingCodeName;
  drg: GroupingCodeName;
  complication: 'MCC' | 'CC' | 'NONE';
  evidence: EvidenceItem[];
  explanation: string;
  candidateRules: CandidateRule[];
  warnings: string[];
}

export interface GroupingError {
  type: string;
  stage: string;
  message: string;
  suggestions: string[];
  candidateMatches?: unknown[];
}

export interface GroupingResultResponse {
  taskId: string;
  status: TaskStatus;
  caseId: string;
  ruleVersion: string;
  startedAt: string;
  finishedAt?: string;
  durationMs?: number;
  result: GroupingResult | null;
  error?: GroupingError;
  inputSnapshot?: Record<string, unknown>;
}

export interface GroupingTaskSummary {
  taskId: string;
  status: TaskStatus;
  caseId: string;
  ruleVersion: string;
  resultDrg?: string;
  startedAt: string;
  finishedAt?: string;
  durationMs?: number;
}
